torch/_functorch/_aot_autograd/subclass_utils.py

```diff
@@ -111,29 +116,117 @@ def create_subclass_meta(
 # a list of tensors that we would then need to concat together.
 # Instead, we specialize the logic for the inference vs. joint graph case.
 # NOTE: this function is hot, since we unwrap tensor subclass inputs at runtime
-def unwrap_tensor_subclasses(wrapped_args, *, is_joint_structure: bool):
-    def concat_inner_tensors_from_subclasses(xs):
-        xs_inner = []
-        for x in xs:
-            if is_traceable_wrapper_subclass(x):
-                xs_inner.extend(get_plain_tensors(typing.cast(Tensor, x)))
-            else:
-                xs_inner.append(x)
-        return xs_inner
 
+
+def filter_symints(size: Iterable[Union[int, SymInt]]) -> List[SymInt]:
+    return [s for s in size if isinstance(s, SymInt)]
+
+
+# The reason for "append_symints"
+#
+# * At compile time: we append extra symint args when unwrapping primals
+# (but not tangents, because they should always share symints with primals).
+# We also append extra symints when unwrapping the subclass outputs of the
+# traced function, so we can return them as extra outputs
+#
+# * At runtime: we similarly append subclass sizes when we unwrap subclass
+# primals (but not tangents) on entry to the forward.
+def unwrap_tensor_subclasses(
+    wrapped_args: List[Union[Tensor, int]],
+    *,
+    is_runtime: bool,
+    append_symints: bool,
+    subclass_metas: Optional[List[Union[int, SubclassCreationMeta]]] = None,
+):
+    xs_inner: List[Union[int, Tensor, SymInt]] = []
+
+    for idx, x in enumerate(wrapped_args):
+        if is_traceable_wrapper_subclass(x):
+            size = x.size()
+            stride = x.stride()
+            xs_inner.extend(get_plain_tensors(typing.cast(Tensor, x)))
+            if append_symints:
+                if is_runtime:
+                    assert subclass_metas is not None
+                    meta = subclass_metas[idx]
+                    assert isinstance(meta, SubclassCreationMeta)
+
+                    # outer_size
+                    symint_placeholders = [type(i) is not int for i in meta.outer_size]
+                    assert len(size) == len(symint_placeholders)
```

Comment:
Why is this guaranteed to be true?

Say I have a tensor subclass where one of its dims is marked dynamic, and the other is not. Then:
(1) `len(x.size()) == 2`
(2) I would expect that `outer_size` looks like `[torch.SymInt, int]`. So `len([type(i) is not int for i in meta.outer_size]) == 1`.

So can we add a test for something like this?
```
@torch.compile(backend="aot_eager_decomp_partition")
def f(x, y):
    return torch.cat([x, y])

x_a = torch.randn(4, 5, 6)
x = TwoTensor(x_a, x_a.clone(), requires_grad=True)

y_a = torch.randn(7, 5, 6)
y = TwoTensor(y_a, y_a.clone(), requires_grad=True)

# dims 0 and 2 are dynamic, dim 1 is static
torch._dynamo.mark_dynamic(x, 0)
torch._dynamo.mark_dynamic(x, 2)
# also mark the inner tensors as dynamic
torch._dynamo.mark_dynamic(x.a, 0)
torch._dynamo.mark_dynamic(x.a, 2)
torch._dynamo.mark_dynamic(x.b, 0)
torch._dynamo.mark_dynamic(x.b, 2)

# dims 0 and 2 are dynamic, dim 1 is static
torch._dynamo.mark_dynamic(y, 0)
torch._dynamo.mark_dynamic(y, 2)
# also mark the inner tensors as dynamic
torch._dynamo.mark_dynamic(y.a, 0)
torch._dynamo.mark_dynamic(y.a, 2)
torch._dynamo.mark_dynamic(y.b, 0)
torch._dynamo.mark_dynamic(y.b, 2)

out = f(x, y)
out.sum().backward()
```

Reply:
> I would expect that outer_size looks like [torch.SymInt, int]. So len([type(i) is not int for i in meta.outer_size]) == 1.

We don't filter the symints out here, but rather just check which indices are not ints. So, len([type(i) is not int for i in meta.outer_size]) == 2, and not 1 as you wrote before.